File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/handlers/EnsureAgentConversationHandler.java
package systems.zlink.samples.supportchat.server.support.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ActorRefSnapshot;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.SupportChannel)
public final class EnsureAgentConversationHandler
        implements ZLinkRequestHandler<
                Messages.EnsureAgentConversationReq, Messages.EnsureAgentConversationRes> {
    private final ZLinkActorManager actors;

    public EnsureAgentConversationHandler(ZLinkActorManager actors) {
        this.actors = actors;
    }

    @Override
    public CompletionStage<Messages.EnsureAgentConversationRes> handle(
            Messages.EnsureAgentConversationReq request, ZLinkMessageContext context) {
        String conversationActorId = request.rosterActorId() + "@" + request.conversationId();
        return actors.find(conversationActorId)
                .thenCompose(
                        existing -> {
                            if (existing.isPresent()) {
                                ActorRef actorRef = existing.orElseThrow();
                                return java.util.concurrent.CompletableFuture.completedFuture(
                                        response(actorRef));
                            }
                            Messages.EnsureSupportUserActorReq create =
                                    new Messages.EnsureSupportUserActorReq(
                                            conversationActorId,
                                            request.displayName(),
                                            SampleNames.Roles.Agent,
                                            request.rosterActorId());
                            return actors.getOrCreate(
                                            conversationActorId, SampleNames.SupportActorType)
                                    .request(create)
                                    .submit()
                                    .thenApply(result -> response(actorRef(result)));
                        });
    }

    private static ActorRef actorRef(ZLinkActorCreateResult result) {
        if (result instanceof ZLinkActorCreateResult.Created created) {
            return created.actor();
        }
        if (result instanceof ZLinkActorCreateResult.Existing existing) {
            return existing.actor();
        }
        throw new IllegalStateException("Agent conversation actor creation was rejected");
    }

    private static Messages.EnsureAgentConversationRes response(ActorRef actorRef) {
        return new Messages.EnsureAgentConversationRes(ActorRefSnapshot.from(actorRef));
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/handlers/EnsureAgentConversationHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.handlers

import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.actors.ActorRef
import systems.zlink.framework.actors.ActorRefSnapshot
import systems.zlink.framework.actors.ZLinkActorClient
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleNames
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureAgentConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureAgentConversationRes
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureSupportUserActorReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.JoinConversationRes

@ZLinkHandlerGroup(SampleNames.SupportChannel)
class EnsureAgentConversationHandler(
    private val actorClient: ZLinkActorClient,
    private val actors: ZLinkActorManager,
) : ZLinkSuspendingRequestHandler<EnsureAgentConversationReq, EnsureAgentConversationRes> {
    override suspend fun handle(
        request: EnsureAgentConversationReq,
        context: ZLinkMessageContext,
    ): EnsureAgentConversationRes {
        val conversationActorId = "${request.rosterActorId}@${request.conversationId}"
        val existingActorRef = actors.find(conversationActorId).await().orElse(null)
        val zlinkActorRef =
            existingActorRef
                ?: actors
                    .kotlin()
                    .getOrCreate(conversationActorId, SampleNames.SupportActorType)
                    .request(
                        EnsureSupportUserActorReq(
                            actorId = conversationActorId,
                            displayName = request.displayName,
                            role = SupportChatRoles.Agent,
                            participantId = request.rosterActorId,
                        )
                    )
                    .await()
                    .requireActor()
        val actorRef: ActorRef = zlinkActorRef

        val joined =
            actorClient
                .requestToActor(
                    actorRef.actorId(),
                    JoinConversationReq(
                        request.rosterActorId,
                        SupportChatRoles.Agent,
                        request.displayName,
                    ),
                )
                .metadata(SampleNames.ConversationIdMetadataKey, request.conversationId)
                .timeout(SampleTimings.RequestTimeout)
                .submit(JoinConversationRes::class.java)
                .await()

        logger.info(
            "support agent conversation: joined. conversation={}, roster={}",
            request.conversationId,
            request.rosterActorId,
        )
        return EnsureAgentConversationRes(
            ActorRefSnapshot.from(actorRef),
            joined.scheduled,
            joined.state,
        )
    }

    private fun ZLinkActorCreateResult.requireActor(): ActorRef =
        when (this) {
            is ZLinkActorCreateResult.Created -> actor
            is ZLinkActorCreateResult.Existing -> actor
            is ZLinkActorCreateResult.Rejected ->
                error("Agent conversation actor creation was rejected")
        }

    private companion object {
        private val logger = LoggerFactory.getLogger(EnsureAgentConversationHandler::class.java)
    }
}
